hyperv: Replace parseLines/SplitLines with getFirstLine

cfergeau opened the pull request:
In the hyperv code, there are many occurrences of resp :=
os.SplitLines(stdout) if resp[0] == ...

This does not check if `resp` is empty before trying to reference it, which
recently triggered a panic in crc ( 
https://github.com/crc-org/crc/issues/3764 )

This code adds a `getFirstLine` helper which will return an empty string if
stdout is "".

This should fix https://github.com/crc-org/crc/issues/3764

cfergeau committed:
Move 'contains' to a new 'strings' package

Its code is duplicated in 2 places with different argument order.

praveenkumar commented:
/lgtm
/approve

gbraad commented on pkg/crc/cluster/clusteroperator.go:
```diff
@@ -101,7 +102,7 @@ func getStatus(ctx context.Context, lister operatorLister, selector []string) (*
 	found := false
 	for _, c := range co.Items {
-		if len(selector) > 0 && !contains(c.ObjectMeta.Name, selector) {
+		if len(selector) > 0 && !crcos.Contains(selector, c.ObjectMeta.Name) {
```
Why do we invert the selector and name?

cfergeau replied:
The function prototype in `clusteroperator.go` was `func contains(value string, list []string) bool`, the new helper is ` Contains(s []string, e string) bool`

gbraad replied:
Thanks for confirming, which means the order has changed. Using more clear variables on the function might make this clearer; `s` is `strings`, and `e` is expected?

This is not the only usage: https://github.com/crc-org/crc/pull/3765/files#r1270660861

gbraad commented on pkg/drivers/hyperv/powershell_windows.go:
```diff
@@ -26,15 +25,13 @@ func cmd(args ...string) error {
-	s := bufio.NewScanner(strings.NewReader(stdout))
-	for s.Scan() {
-		resp = append(resp, s.Text())
+func getFirstLine(stdout string) string {
```
Not part of the util.go ?

cfergeau replied:
This package is the only user of `getFirstLine` so I kept it there as it's a bit specialized.

gbraad replied:
the issue in question was the only place for `resp[0]`, for which we also use the `| get-unique` operator in powershell. I don't mind, but found it very generic.

cfergeau replied:
ok I moved it.

gbraad commented on pkg/os/util.go:
```diff
@@ -113,3 +114,24 @@ func RemoveFileIfExists(path string) error {
 }
+
+// Split a multi line string in an array of string, one for each line
+func SplitLines(input string) []string {
```
why is this part of crcos (operating system-specific code). this looks like generic line/string manipulation

cfergeau replied:
I could not find a more appropriate package, and was not sure adding a `strings` package for 2 functions was worh it. I can make this change if you prefer.

gbraad replied:
I think adding a strings make sense, as I might even think adding the firstline justifies the use. it makes it clearer in intent. perhaps we might even add a getlastline at some point ;-)

cfergeau replied:
Done.

gbraad commented on pkg/crc/services/dns/dns_windows.go:
```diff
@@ -30,7 +29,7 @@ func runPostStartForOS(serviceConfig services.ServicePostStartConfig) error {
 	time.Sleep(2 * time.Second)
 
-	if !contains(getInterfaceNameserverValues(networkInterface), serviceConfig.IP) {
+	if !crcos.Contains(getInterfaceNameserverValues(networkInterface), serviceConfig.IP) {
```
then is this not in error?

```
if !crcos.Contains(getInterfaceNameserverValues(networkInterface), serviceConfig.IP) {
```

the order for the parameters have changed? https://github.com/crc-org/crc/pull/3765/files#r1270651088 Let's stay consistent.

cfergeau replied:
`dns_windows.go` was using `contains(s []string, e string) bool` and `clusteroperator.go` used `contains(value string, list []string) bool`,  so the order will change in one of the instances.

gbraad replied:
Check

Can we in that case also more clearly describe the parameters:

```
func Contains(s []string, e string) bool {
```

with actual 'words' instead of abbreviated identifiers. https://github.com/crc-org/crc/pull/3765#discussion_r1270661563

`s` = `strings` and `e` = `expected`?

gbraad commented on pkg/strings/strings.go:
```diff
@@ -0,0 +1,36 @@
+	return output
+}
+
+func FirstLine(stdout string) string {
```
`s/stdOut/input/`, just like in `SplitLines`. `stdOut` does not have any meaning here, except from the caller.

gbraad commented on pkg/strings/strings.go:
```diff
@@ -0,0 +1,36 @@
+	"strings"
+)
+
+func Contains(s []string, e string) bool {
```
`s/s/strings/`, `s/e/expected/` or `compare`. other functions use more descriptive parameters, like `input`.

cfergeau replied:
I used needle/haystack, I can change again if this is harder to read than it should.

gbraad reviewed (changes requested):
parameters are not clear/consistent.

otherwise lgtm

cfergeau commented:
Let's hold this for a little bit, I'll add some unit tests.

cfergeau committed:
Add SplitLines helper to pkg/strings

We have 2 duplicate implementations of this code.
It's renamed to 'Split' instead of 'Parse' to be more consistent with
other go APIs.

cfergeau committed:
hyperv: Replace SplitLines with FirstLine

In the hyperv code, there are many occurrences of
resp := os.SplitLines(stdout)
if resp[0] == ...

This does not check if `resp` is empty before trying to reference it,
which recently triggered a panic in crc (
https://github.com/crc-org/crc/issues/3764 )

This code adds a `FirstLine` helper which will return an empty string
if stdout is "".

This should fix https://github.com/crc-org/crc/issues/3764

cfergeau commented:
I've now added unit tests, and they made me wonder if it would not be best to change `SplitLines("")` to return `{ "" }` instead of `{ }`, as at the moment it's a bit weird that `SplitLines("a\n")` returns `{ "a" }`, `SplitLines("a")` also returns `{ "a" }`, but we don't have the same behaviour if we use an empty string instead of `"a"`

gbraad commented:
why haystack and needle? stay consistent with the rest and use `input` and somehing like `match`. As in a `findNeedle`-like function I expect the identifier to be returned of the `needle` that matches, not just a `bool`-value.

gbraad commented on pkg/strings/strings.go:
```diff
@@ -0,0 +1,36 @@
+	"strings"
+)
+
+func Contains(haystack []string, needle string) bool {
```
inconsistent with other naming; also, this does not work like a `FindNeedle`-function.

anjannath commented:
> I've now added unit tests, and they made me wonder if it would not be best to change `SplitLines("")` to return `{ "" }` instead of `{ }`, as at the moment it's a bit weird that `SplitLines("a\n")` returns `{ "a" }`, `SplitLines("a")` also returns `{ "a" }`, but we don't have the same behaviour if we use an empty string instead of `"a"`

with just `"\n"` its an empty line but with `""` there's no line, so maybe not returning `{ "" }` is better or we could decide to treat empty lines as no line and return `{}` for `"\n"` this will make it consistent

cfergeau commented:
> > I've now added unit tests, and they made me wonder if it would not be best to change `SplitLines("")` to return `{ "" }` instead of `{ }`, as at the moment it's a bit weird that `SplitLines("a\n")` returns `{ "a" }`, `SplitLines("a")` also returns `{ "a" }`, but we don't have the same behaviour if we use an empty string instead of `"a"`
> 
> with just `"\n"` its an empty line but with `""` there's no line, so maybe not returning `{ "" }` is better or we could decide to treat empty lines as no line and return `{}` for `"\n"` this will make it consistent

The initial bug is that `{}` is not handled correctly by our code, if we never return `{}`, there is no bug :)
`{}` for `"\n'` would not be that consistent, as `"\n\n"` returns `{ "", ""}`
